mllib/src/main/scala/org/apache/spark/ml/feature/StringIndexer.scala

```diff
@@ -107,11 +135,15 @@ class StringIndexer @Since("1.4.0") (
   @Since("2.0.0")
   override def fit(dataset: Dataset[_]): StringIndexerModel = {
     transformSchema(dataset.schema, logging = true)
-    val counts = dataset.na.drop(Array($(inputCol))).select(col($(inputCol)).cast(StringType))
-      .rdd
-      .map(_.getString(0))
-      .countByValue()
-    val labels = counts.toSeq.sortBy(-_._2).map(_._1).toArray
+    val values = dataset.na.drop(Array($(inputCol)))
+      .select(col($(inputCol)).cast(StringType))
+      .rdd.map(_.getString(0))
+    val labels = $(stringOrderType) match {
+      case "freq_desc" => values.countByValue().toSeq.sortBy(-_._2).map(_._1).toArray
```

Comment:
Seems this setting is case-insensitive because your param validator doesn't care it, `$(stringOrderType)` might be upper-case here. We may make it case-sensitive or do `toLowerCase` here?